File: mobile/AnkidoteMobile/AnkidoteMobile/Loop/IRT.swift
import Foundation

/// Item Response Theory primitives — a direct port of `pylib/anki/ankidote/irt.py`
/// and `scores.py` so the mobile loop re-estimates ability and score bands
/// identically to the desktop.
enum IRT {
    /// Logistic scaling constant (normal-ogive scale), matching catsim's default.
    static let D = 1.702
    static let thetaMin = -4.0
    static let thetaMax = 4.0
    static let maxSE = 1.5

    struct Params {
        let a: Double
        let b: Double
        let c: Double
    }

    /// 3PL probability of a correct response at the given ability.
    static func pCorrect(theta: Double, _ p: Params) -> Double {
        p.c + (1.0 - p.c) / (1.0 + exp(-D * p.a * (theta - p.b)))
    }

    /// Fisher information contributed by an item at `theta` (3PL).
    static func information(theta: Double, _ p: Params) -> Double {
        let prob = pCorrect(theta: theta, p)
        if prob <= p.c || prob >= 1.0 { return 0.0 }
        let quotient = (prob - p.c) / (1.0 - p.c)
        return pow(D * p.a, 2) * (quotient * quotient) * (1.0 - prob) / prob
    }

    private static func logLikelihood(theta: Double, params: [Params], responses: [Bool]) -> Double {
        var total = 0.0
        for (p, correct) in zip(params, responses) {
            let prob = min(max(pCorrect(theta: theta, p), 1e-9), 1.0 - 1e-9)
            total += correct ? log(prob) : log(1.0 - prob)
        }
        return total
    }

    /// Maximum-likelihood ability estimate via coarse-then-fine grid search.
    static func estimateTheta(params: [Params], responses: [Bool], prior: Double = 0.0) -> Double {
        if responses.isEmpty { return prior }

        if Set(responses).count == 1 {
            // Monotonic likelihood → pin toward the evidence, nudged by prior.
            let direction = responses[0] ? 1.0 : -1.0
            return max(thetaMin, min(thetaMax, prior + direction * 1.5))
        }

        var bestTheta = prior
        var bestLL = -Double.infinity
        // (scale, span): coarse full-range pass, then a fine refine around best.
        let passes: [(scale: Double, span: Double?)] = [(0.05, nil), (0.005, 0.1)]
        for pass in passes {
            let lo = pass.span == nil ? thetaMin : max(thetaMin, bestTheta - pass.span!)
            let hi = pass.span == nil ? thetaMax : min(thetaMax, bestTheta + pass.span!)
            let steps = Int(((hi - lo) / pass.scale).rounded())
            for i in 0 ... max(steps, 0) {
                let theta = lo + Double(i) * pass.scale
                let ll = logLikelihood(theta: theta, params: params, responses: responses)
                if ll > bestLL {
                    bestLL = ll
                    bestTheta = theta
                }
            }
        }
        return bestTheta
    }

    /// Standard error of the ability estimate = 1/sqrt(test information).
    static func standardError(theta: Double, params: [Params]) -> Double {
        let totalInfo = params.reduce(0.0) { $0 + information(theta: theta, $1) }
        if totalInfo <= 0.0 { return maxSE }
        return min(1.0 / sqrt(totalInfo), maxSE)
    }
}

/// theta/SE → GMAT score-range mapping (port of `scores.py`).
enum Scores {
    static let scoreMid = 505.0
    static let scorePerTheta = 100.0
    static let scoreLow = 205
    static let scoreHigh = 805
    static let scoreStep = 10
    static let z = 1.96

    static func thetaToScore(_ theta: Double) -> Int {
        let raw = scoreMid + theta * scorePerTheta
        let clamped = min(max(raw, Double(scoreLow)), Double(scoreHigh))
        let steps = Int(((clamped - Double(scoreLow)) / Double(scoreStep)).rounded())
        let maxSteps = (scoreHigh - scoreLow) / scoreStep
        return scoreLow + min(steps, maxSteps) * scoreStep
    }

    static func scoreRange(theta: Double, se: Double) -> ScoreRange {
        ScoreRange(low: thetaToScore(theta - z * se), high: thetaToScore(theta + z * se))
    }

    /// Combine per-topic (theta, se, weight) into overall (theta, se).
    static func combineTopics(_ estimates: [(theta: Double, se: Double, weight: Double)]) -> (theta: Double, se: Double) {
        let weighted = estimates.filter { $0.weight > 0 }
        if weighted.isEmpty { return (0.0, 1.5) }
        let totalW = weighted.reduce(0.0) { $0 + $1.weight }
        let theta = weighted.reduce(0.0) { $0 + $1.theta * $1.weight } / totalW
        let variance = weighted.reduce(0.0) { $0 + pow($1.weight / totalW, 2) * $1.se * $1.se }
        return (theta, sqrt(variance))
    }
}

File: pylib/anki/ankidote/irt.py
# Copyright: Ankitects Pty Ltd and contributors
# License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

"""Item Response Theory primitives for the Ankidote diagnostic.

This is the mathematical core the CAT engine is built on: the 3PL item
characteristic curve, Fisher information (for max-information item
selection), a maximum-likelihood ability estimator, and the standard error
of that estimate.

It is a deliberately small, dependency-free implementation for the
temp-question milestone. The diagnostic-cat-plan calls for `catsim` at
runtime; the functions here mirror the pieces of `catsim.irt` we rely on
(`icc`, `inf`, an estimator, and `see`) so that swapping in `catsim` later is
a localized change behind ``CatSession`` — nothing else in the app needs to
know which implementation produced theta/SE.
"""

from __future__ import annotations

import math
from collections.abc import Sequence

# Logistic scaling constant that puts the logistic model on (approximately)
# the normal-ogive scale, matching catsim's default.
D = 1.702

# Ability search bounds. GMAT abilities are mapped onto a standard-normal
# theta scale, so [-4, 4] covers virtually the whole population.
THETA_MIN = -4.0
THETA_MAX = 4.0

# Standard error reported before any information has been collected, and the
# ceiling we clamp SE to so an empty/near-empty response set does not produce
# an absurd score band.
MAX_SE = 1.5


def p_correct(theta: float, a: float, b: float, c: float) -> float:
    """3PL probability of a correct response at the given ability."""
    return c + (1.0 - c) / (1.0 + math.exp(-D * a * (theta - b)))


def information(theta: float, a: float, b: float, c: float) -> float:
    """Fisher information contributed by an item at ``theta`` (3PL)."""
    p = p_correct(theta, a, b, c)
    if p <= c or p >= 1.0:
        return 0.0
    quotient = (p - c) / (1.0 - c)
    return (D * a) ** 2 * (quotient**2) * (1.0 - p) / p


def _log_likelihood(
    theta: float,
    params: Sequence[tuple[float, float, float]],
    responses: Sequence[float],
) -> float:
    """Log-likelihood for graded responses ``g ∈ [0, 1]``.

    A dichotomous response (``0``/``1`` or ``bool``) reduces to the usual
    ``log P`` / ``log(1−P)``. Partial credit ``g`` (e.g. from answer-choice
    ranking) contributes ``g·log P + (1−g)·log(1−P)`` — the cross-entropy of the
    graded score against the item characteristic curve.
    """
    total = 0.0
    for (a, b, c), g in zip(params, responses):
        p = min(max(p_correct(theta, a, b, c), 1e-9), 1.0 - 1e-9)
        g = min(max(float(g), 0.0), 1.0)
        total += g * math.log(p) + (1.0 - g) * math.log(1.0 - p)
    return total


def estimate_theta(
    params: Sequence[tuple[float, float, float]],
    responses: Sequence[float],
    *,
    prior: float = 0.0,
) -> float:
    """Maximum-likelihood ability estimate via a coarse-then-fine grid search.

    ``responses`` may be dichotomous (``bool``/``0``/``1``) or graded floats in
    ``[0, 1]`` (partial credit). ``prior`` (the survey-seeded warm start) is
    returned unchanged when there is no response yet, and breaks ties for
    all-extreme response patterns, whose MLE is otherwise unbounded.
    """
    if not responses:
        return prior

    numeric = [min(max(float(g), 0.0), 1.0) for g in responses]
    all_extreme = all(g <= 1e-9 for g in numeric) or all(
        g >= 1.0 - 1e-9 for g in numeric
    )
    if all_extreme:
        # Likelihood is monotonic; the MLE runs off to ±infinity. Pin the
        # estimate to a sensible bound in the direction of the evidence,
        # nudged by the prior, rather than the grid extreme.
        direction = 1.0 if numeric[0] >= 0.5 else -1.0
        return max(THETA_MIN, min(THETA_MAX, prior + direction * 1.5))
    responses = numeric

    best_theta = prior
    best_ll = -math.inf
    # Coarse pass over the whole range, then refine around the best point.
    for scale, center, span in ((0.05, 0.0, None), (0.005, None, 0.1)):
        lo = THETA_MIN if span is None else max(THETA_MIN, best_theta - span)
        hi = THETA_MAX if span is None else min(THETA_MAX, best_theta + span)
        steps = int(round((hi - lo) / scale))
        for i in range(steps + 1):
            theta = lo + i * scale
            ll = _log_likelihood(theta, params, responses)
            if ll > best_ll:
                best_ll = ll
                best_theta = theta
    return best_theta


def standard_error(theta: float, params: Sequence[tuple[float, float, float]]) -> float:
    """Standard error of the ability estimate = 1/sqrt(test information)."""
    total_info = sum(information(theta, a, b, c) for a, b, c in params)
    if total_info <= 0.0:
        return MAX_SE
    return min(1.0 / math.sqrt(total_info), MAX_SE)
